File: codechef/dishown.cpp
#include <bits/stdc++.h>
using namespace std;

int main(){
    

    int tt;
    cin >> tt;
    while(tt--){
        int n; cin >> n;
        vector <int> scores(n);
        for(int i = 0; i < n; i++) cin >> scores[i];
        vector <int> owner(n);
        for(int i = 0; i < n; i++) owner[i] = i+1;
        int q;
        cin >> q;
        while(q--){
            int x; cin >> x;
            if(x == 1){
                int y;
                cin >> y;
                cout << owner[y-1] << endl;
            }
            else{
                int y, z;
                cin >> y >> z;
                if(owner[y-1] == owner[z-1]) cout << "Invalid query!" << endl;
                if(scores[y-1] > scores[z-1]) {
                    int loser = owner[z-1];
                    for(int i = 0; i < n; i++){
                        if(owner[i] == loser) owner[i] = owner[y-1];
                    }
                }
                else if(scores[y-1] < scores[z-1]) {
                    int loser = owner[y-1];
                    for(int i = 0; i < n; i++){
                        if(owner[i] == loser) owner[i] = owner[z-1];
                    }
                    
                }
            }
        }
    }
}

File: codechef/dishown.py
tt = int(input())
for _ in range(tt):
    n = int(input())
    scores = [int(s) for s in input().split()]
    owner = list(range(1, n+1))
    q = int(input())
    for qq in range(q):
        inp = [int(s) for s in input().split()]
        if inp[0] == 1:
            print(owner[inp[1]-1])
        else:
            if owner[inp[1]-1] == owner[inp[2]-1]:
                print("Invalid query!")
            elif scores[inp[1]-1] > scores[inp[2]-1]:
                owner[inp[2]-1] = owner[inp[1]-1]
            elif scores[inp[1]-1] < scores[inp[2]-1]:
                owner[inp[1]-1] = owner[inp[2]-1]
            else:
                pass
            
